### ISSUE TITLE
StackOverflow when modifying the FQN of an aspect in a `with` expression

### ISSUE THREAD

**Issue by ebousse** [diverse-project/melange#96]
Versions:
- Melange 0.2.1.201710121221
- Xtext 2.12.0.v20170519-0902
- EMF Ecore 2.13.0.v20170609-0707

When editing a Melange file and introducing a space in the FQN of an aspect, things get instantly ugly and Eclipse must close, as shown here:

![capture d ecran de 2017-11-08 13-07-56](https://user-images.githubusercontent.com/5868014/32548445-f877d522-c485-11e7-82d4-e83a288b83a1.png)

When looking at the error log, we see two errors:
1. « An internal error occurred during: "Xtext validation". »
2. « An internal error occurred during: "XtextReconcilerJob". »

Both errors are triggered by a `java.lang.StackOverflowError`, and both because EMF is stuck in a recursion when trying to resolve the proxy of the EClass found in the melange fragment corresponding to an aspect.

Here is a sample of the stack trace, where the first and last lines are identical, and more precisely where the two calls to `resolve` are identical (same arguments) as can be observed with a debugger:
```
...
	at org.eclipse.emf.ecore.util.EcoreUtil.resolve(EcoreUtil.java:261)
	at org.eclipse.emf.ecore.impl.BasicEObjectImpl.eResolveProxy(BasicEObjectImpl.java:1477)
	at org.eclipse.emf.ecore.util.EcoreEList.resolveProxy(EcoreEList.java:206)
	at org.eclipse.emf.ecore.util.EcoreEList.resolve(EcoreEList.java:161)
	at org.eclipse.emf.ecore.util.EObjectContainmentWithInverseEList$Resolving.resolve(EObjectContainmentWithInverseEList.java:105)
	at org.eclipse.emf.common.util.BasicEList.get(BasicEList.java:348)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.doNext(AbstractEList.java:707)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.next(AbstractEList.java:694)
	at org.eclipse.emf.ecore.impl.EPackageImpl.getEClassifierGen(EPackageImpl.java:446)
	at org.eclipse.emf.ecore.impl.EPackageImpl.eObjectForURIFragmentSegment(EPackageImpl.java:1873)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.getEObject(ResourceImpl.java:766)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.getEObject(ResourceImpl.java:742)
	at org.eclipse.xtext.resource.XtextResource.access$101(XtextResource.java:63)
	at org.eclipse.xtext.resource.XtextResource$1.getEObject(XtextResource.java:119)
	at org.eclipse.xtext.resource.DefaultFragmentProvider.getEObject(DefaultFragmentProvider.java:28)
	at org.eclipse.xtext.resource.XtextResource.basicGetEObject(XtextResource.java:363)
	at org.eclipse.xtext.resource.XtextResource.getEObject(XtextResource.java:349)
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.getEObject(LazyLinkingResource.java:232)
	at org.eclipse.xtext.resource.persistence.StorageAwareResource.getEObject(StorageAwareResource.java:120)
	at org.eclipse.xtext.xbase.resource.BatchLinkableResource.getEObject(BatchLinkableResource.java:124)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.getEObject(ResourceSetImpl.java:223)
	at org.eclipse.emf.ecore.util.EcoreUtil.resolve(EcoreUtil.java:201)
	at org.eclipse.emf.ecore.util.EcoreUtil.resolve(EcoreUtil.java:261)
...
```

Note that while the error sometimes happens when Melange is explicitly doing some pre(or post)-processing, and can hence be prevented by introducing a check there, other errors are automatically triggered by Xtext, hence it never goes through Melange code. 

Hence I have no idea where should the fix be made: in EMF? in Xtext? when making the fragments in Melange? unclear and very hard to understand the exact cause...

## To reproduce

Set up the workspace:
- Create a melange project
- Create a melange file with the content:
```
package X
language Test {

}
```
- In the language declaration, use Ctrl+Space to "Create a Domain Model Project"
- Create 2 classes `A` and `B` in the ecore file
- Generate code with genmodel
- Back in the language declaration, use Ctrl+Space to "Create a DSA Project" from the ecore file
- The melange file now contains
```
package X

language Test {
	syntax "platform:/resource/x.test.model/model/model.ecore"
	with model.aspects.BAspect
	with model.aspects.AAspect
}
```

Trigger the bug:
- Click to put your text cursor in the FQN of one of the aspect (eg. `model.aspects.AAspect`)
- Press SPACE to add a space in the middle of the FQN

**Comment by tdegueul** [comment]
@ebousse, thank you so much for the detailed report!

I've also encountered this exact error recently but was not able to find a reliable way to trigger it, so this helps a lot!

I'm also a bit puzzled by the fact that the root error doesn't go through Melange itself. I dived into it using the debugger but this still does not make any sense to me.

Pinging @fcoulon just in case. I'll find some time to dive into it again, because this is critical.

**Comment by ebousse** [comment]
I have shown this to @fcoulon already, and it's really hard to get what's going on x_x

Also here I'm only mentioning the case where a typo is introduced in the Aspect FQN, but I'm pretty sure the problem occurs in other cases that I have not identified as precisely as this one.

If anyone can confirm that the error happens on his computer with the scenario I have described, that would confirm that we have at least one precise reproducible scenario.


And unfortunately I agree that this is critical. Melange is almost completely unusable because of these crashes :(.

**Comment by fcoulon** [comment]
No more StackOverflow if I comment this line:
https://github.com/diverse-project/melange/blob/master/plugins/fr.inria.diverse.melange/src/main/java/fr/inria/diverse/melange/processors/LanguageProcessor.xtend#L180

Need more investigation :/

**Comment by ebousse** [comment]
But this line seems important, I guess that commenting it breaks Melange?

**Comment by fcoulon** [comment]
Seems related to https://github.com/eclipse/xtext/issues/1121

**Comment by ebousse** [comment]
@fcoulon yes that's definitely exactly the same issue. A workaround is given in this comment in the discussion thread: https://github.com/eclipse/xtext/issues/1121#issuecomment-312717933 . Worth a try?

**Comment by fcoulon** [comment]
Yep I tried 
~~~
def Class<? extends IReferableElementsUnloader> bindIReferableElementsUnloader() {
		return IReferableElementsUnloader.NullUnloader
	}
~~~
and there is no more StackOverflow

**Comment by ebousse** [comment]
Hurray then :)

**Comment by tdegueul** [comment]
Also tried locally, no more SO.

Can you push the "fix" @fcoulon?

Thank you guys.

**Comment by fcoulon** [comment]
Should be fixed by ab0c342cc5ef5b8a992351c295c99b3fede77ed7

**Comment by ebousse** [comment]
I can confirm as well that this works. Great job @fcoulon.
